lib/mayaUsd/ufe/UsdUndoConnectionCommands.cpp

```diff
@@ -284,6 +284,139 @@ UsdUndoDeleteConnectionCommand::Ptr UsdUndoDeleteConnectionCommand::create(
     return std::make_shared<UsdUndoDeleteConnectionCommand>(srcAttr, dstAttr);
 }
 
+bool canRemoveSrcProperty(const PXR_NS::UsdAttribute& srcAttr)
+{
+
+    // Do not remove if it has a value.
+    if (srcAttr.HasValue()) {
+        return false;
+    }
+
+    PXR_NS::SdfPathVector connectedAttrs;
+    srcAttr.GetConnections(&connectedAttrs);
+
+    // Do not remove if it has connections.
+    if (!connectedAttrs.empty()) {
+        return false;
+    }
+
+    const auto prim = srcAttr.GetPrim();
+
+    if (!prim) {
+        return false;
+    }
+
+    const auto primParent = prim.GetParent();
+
+    if (!primParent) {
+        return false;
+    }
+
+    // Do not remove if there is a connection with a prim.
+    for (const auto& childPrim : primParent.GetChildren()) {
+        if (childPrim != prim) {
+            for (const auto& attribute : childPrim.GetAttributes()) {
+                const PXR_NS::UsdAttribute dstUsdAttr = attribute.As<PXR_NS::UsdAttribute>();
+                if (isConnected(srcAttr, dstUsdAttr)) {
+                    return false;
+                }
+            }
+        }
+    }
+
+    // Do not remove if there is a connection with the parent prim.
+    for (const auto& attribute : primParent.GetAttributes()) {
+        const PXR_NS::UsdAttribute dstUsdAttr = attribute.As<PXR_NS::UsdAttribute>();
+        if (isConnected(srcAttr, dstUsdAttr)) {
+            return false;
+        }
+    }
+
+    PXR_NS::UsdShadeNodeGraph ngPrim(prim);
+
+    if (!ngPrim) {
+        return true;
+    }
+
+    // Do not remove if there is a connection with a child prim.
+    for (const auto& childPrim : prim.GetChildren()) {
+        for (const auto& attribute : childPrim.GetAttributes()) {
+            const PXR_NS::UsdAttribute dstUsdAttr = attribute.As<PXR_NS::UsdAttribute>();
+            if (isConnected(srcAttr, dstUsdAttr)) {
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+bool canRemoveDstProperty(const PXR_NS::UsdAttribute& srcAttr)
+{
+    // Do not remove if it has a value.
+    if (srcAttr.HasValue()) {
+        return false;
+    }
+
+    PXR_NS::SdfPathVector connectedAttrs;
+    srcAttr.GetConnections(&connectedAttrs);
+
+    // Do not remove if it has connections.
+    if (!connectedAttrs.empty()) {
+        return false;
+    }
+
+    const auto prim = srcAttr.GetPrim();
+
+    if (!prim) {
+        return false;
+    }
+
+    PXR_NS::UsdShadeNodeGraph ngPrim(prim);
+
+    if (!ngPrim) {
+        return true;
```

Comment:
the functions `canRemoveSrcProperty `and `canRemoveDstProperty `are slightly different: if the prim is not a `PXR_NS::UsdShadeNodeGraph` and the attribute is a destination, we can return true immediately, instead if the attribute were a source, we have to do more checks.